include/real/exact_number.hpp

```diff
@@ -264,6 +272,116 @@ namespace boost {
                 this->normalize();
             }
 
+            /**     KARATSUBA MULTIPLICATION
+             *  @brief: multiplies (*this) with other using karatsuba multiplication algorithm
+             *  @param: other: an exact_number to be multiplied with (*this)
+             *  @param: base: base of the numbers being multiplied
+             *  @author: Kishan Shukla
+             */
+
+            void karatsuba_multiplication (
+                    exact_number<T> &other, 
+                    const T base = (std::numeric_limits<T>::max() / 4) * 2
+            ) {
+
+                while (this->exponent > (int)this->digits.size()) {
+                    this->digits.push_back(0);
+                }
+
+                while (other.exponent > (int)other.digits.size()) {
+                    other.digits.push_back(0);
+                }
+
+                // this --- a, other --- b
+
+            	int a_size = this->digits.size();
+            	int b_size = other.digits.size();
+
+                const int max_length = std::max(a_size, b_size);
+
+                if (max_length <= KARATSUBA_THRESHOLD) {
+                    this->standard_multiplication(other, base);
+                    return;
+                }
+
+
+                // appending zeroes in front to make sizes of a & b equal
+                std::vector<T> final_a, final_b;
+                if (a_size < max_length) {
+                	std::vector<T> zeroes(max_length - a_size, 0);
+                	final_a.reserve(max_length);
+                	final_a.insert(final_a.end(), zeroes.begin(), zeroes.end());
+                	final_a.insert(final_a.end(), this->digits.begin(), this->digits.end());
+                	final_b = other.digits;
```

Comment:
Isn't this potentially to expensive of an operation compared to passing around the responsability of implicitly knowing them are 0s?
For example, you could use a wrapper class for vector where the operator[] returns zero for the X first elements and redirects to the right digits in a when neeeded. Additionally you would need begin, end, and operator*.
That will save you the reservation of the new vector, the waste on digits that are 0s at left and the "copy" from a.

Reply:
yes, it is expensive. I didn't earlier notice that this can be removed. I have changed this and there was no need for a wrapper class as this can be easily removed by just storing how many zeroes I am going to add. I have made the necessary changes.